app/src/org/commcare/views/widgets/MediaWidget.java

```diff
@@ -162,4 +180,117 @@ public void setFocus(Context context) {
+        }
+
+        File newMedia;
+
+        if (!destMediaPath.isEmpty()) {
+            // we have already copied the file at newPath during createFilePath
+            newMedia = new File(destMediaPath);
+        } else {
+            recordedFileName = FileUtil.getFileName(binaryPath);
+            destMediaPath = mInstanceFolder + System.currentTimeMillis() + customFileTag + FileUtil.getExtension(binaryPath);
+
+            // Copy to destMediaPath
+            File source = new File(binaryPath);
+            newMedia = new File(destMediaPath);
+            try {
+                FileUtil.copyFile(source, newMedia);
+            } catch (IOException e) {
+                showToast("form.attachment.copy.fail");
+                Log.e(TAG, "IOExeception while copying media");
+                e.printStackTrace();
+            }
+        }
+
+        if (newMedia.exists()) {
+            // Add the copy to the content provider
+            ContentValues values = new ContentValues(6);
+            values.put(MediaStore.Audio.Media.TITLE, newMedia.getName());
+            values.put(MediaStore.Audio.Media.DISPLAY_NAME, newMedia.getName());
+            values.put(MediaStore.Audio.Media.DATE_ADDED, System.currentTimeMillis());
+            values.put(MediaStore.Audio.Media.DATA, newMedia.getAbsolutePath());
+
+            Uri mediaUri =
+                    getContext().getContentResolver().insert(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, values);
+            String mediaUriString = mediaUri == null ? "null" : mediaUri.toString();
+            Log.i(TAG, "Inserting media returned uri = " + mediaUriString);
+            showToast("form.attachment.success");
+        } else {
+            Log.e(TAG, "Inserting media file FAILED");
+        }
+
+        mBinaryName = newMedia.getName();
+    }
+
+    /**
+     * If file is chosen by user, the file selection intent will return an URI
+     * If file is auto-selected after recording_fragment, then the recordingfragment will provide a string file path
+     * Set value of customFileTag if the file is a recent recording from the RecordingFragment
+     */
+    protected String createFilePath(Object binaryuri) {
+        String path = "";
+        destMediaPath = "";
+        if (binaryuri instanceof Uri) {
+            try {
+                path = UriToFilePath.getPathFromUri(CommCareApplication.instance(),
+                        (Uri)binaryuri);
+            } catch (UriToFilePath.NoDataColumnForUriException e) {
+                // Need to make a copy of file using uri, so might as well copy to final destination path directly
+                InputStream inputStream;
+                try {
+                    inputStream = getContext().getContentResolver().openInputStream((Uri)binaryuri);
+                } catch (FileNotFoundException e1) {
+                    showToast("form.attachment.notfound");
+                    e1.printStackTrace();
+                    return "";
```

Comment:
Same question as above WRT whether we shouldn't be more deliberate here with Exceptions rather than potentially ambiguous return types.